src/server/pfs/server/driver.go

```diff
@@ -806,188 +700,116 @@ func (d *driver) repoSize(ctx context.Context, repo *pfs.Repo) (int64, error) {
+		return len(propagatedBranches[i].Provenance) < len(propagatedBranches[j].Provenance)
 	})
-
-	// Iterate through downstream branches and add new commits.
-	hasNewCommits := false
-	for _, subvBI := range subvBIs {
+	// add new commits, set their ancestry + provenance pointers, and advance branch heads
+	for _, bi := range propagatedBranches {
+		// TODO(acohen4): can we just make calls to startCommit() here?
 		// Do not propagate an open commit onto spout output branches (which should
 		// only have a single provenance on a spec commit)
-		if len(subvBI.Provenance) == 1 && subvBI.Provenance[0].Repo.Type == pfs.SpecRepoType {
+		if len(bi.Provenance) == 1 && bi.Provenance[0].Repo.Type == pfs.SpecRepoType {
 			continue
 		}
-		hasNewCommits = true
-
-		// Create aliases for any provenant branches which are not already part of this CommitSet
-		for _, provOfSubvB := range subvBI.Provenance {
-			provOfSubvBI, err := getBranchInfo(provOfSubvB)
-			if err != nil {
-				return err
-			}
-			if provOfSubvBI.Head.ID != txnCtx.CommitSetID {
-				if _, err := d.aliasCommit(txnCtx, provOfSubvBI.Head, provOfSubvBI.Head.Branch); err != nil {
-					return err
-				}
-				// Update the cached branch head
-				provOfSubvBI.Head.ID = txnCtx.CommitSetID
-			}
-			// if this is a pipeline output branch, we need to also create an alias commit on the meta branch
-			// to maintain pipeline system invariants.
-			if provOfSubvBI.Branch.Repo.Type == pfs.UserRepoType {
-				metaBranch := client.NewSystemProjectRepo(provOfSubvBI.Branch.Repo.Project.GetName(), provOfSubvBI.Branch.Repo.Name, pfs.MetaRepoType).
-					NewBranch(provOfSubvBI.Branch.Name)
-				metaBI, err := getBranchInfo(metaBranch)
-				if err != nil {
-					if col.IsErrNotFound(err) {
-						// no corresponding meta branch, so not a pipeline. Ignore
-						continue
-					}
-					return err
+		// TODO(acohen4): understand why this is necessary
+		if bi.GetHead().GetID() == txnCtx.CommitSetID {
+			continue
+		}
+		newCommit := &pfs.Commit{
+			Repo:   bi.Branch.Repo,
+			Branch: bi.Branch,
+			ID:     txnCtx.CommitSetID,
+		}
+		newCommitInfo := &pfs.CommitInfo{
+			Commit:           newCommit,
+			Origin:           &pfs.CommitOrigin{Kind: pfs.OriginKind_AUTO},
+			Started:          txnCtx.Timestamp,
+			DirectProvenance: bi.DirectProvenance,
+		}
+		// we might be able to find an older parent commit that better reflects the provenance state, saving work
+		// Set 'newCommit's ParentCommit, 'branch.Head's ChildCommits and 'branch.Head'
+		newCommitInfo.ParentCommit = proto.Clone(bi.Head).(*pfs.Commit)
+		bi.Head = newCommit
+		if err := d.branches.ReadWrite(txnCtx.SqlTx).Put(bi.Branch, bi); err != nil {
+			return errors.Wrapf(err, "put branch %q with head %q", pfsdb.BranchKey(bi.Branch), pfsdb.CommitKey(bi.Head))
+		}
+		// create open 'commit'.
+		// it's possible that this commit has already been created if there are two branches with the same head that are propagated.
```

Comment:
I think this goes away if we don't allow branches in the same repo to be in the same provenance DAG.